lib/changesreader.js

```diff
@@ -0,0 +1,284 @@
+  // - batchSize - the number of records to return per HTTP request
+  // - since - the the sequence token to start from (defaults to 'now')
+  start (opts) {
+    const self = this
+
+    // if we're already listening for changes
+    if (self.started) {
+      // return the existing event emitter
+      return self.ee
+    }
+    self.started = true
+
+    // handle overidden defaults
+    opts = opts || {}
+    Object.assign(self, opts)
+
+    // the work function is async and runs in the background
+    // with a big do/while loop
+    const work = async () => {
+      do {
+        // formulate changes feed longpoll HTTP request
+        let pause = 0
+        const req = {
+          method: 'post',
+          db: self.db,
+          path: '_changes',
+          qs: {
+            feed: 'longpoll',
+            timeout: self.timeout,
+            since: self.since,
+            limit: self.batchSize,
+            include_docs: self.includeDocs
+          },
+          body: {}
+        }
+        if (self.fastChanges) {
+          req.qs.seq_interval = self.batchSize
+        }
+        if (self.selector) {
+          req.qs.filter = '_selector'
+          req.body.selector = self.selector
+        }
+        Object.assign(req.qs, opts.qs)
+
+        // make HTTP request to get up to batchSize changes from the feed
+        try {
+          const data = await self.request(req)
+
+          // update the since state
+          if (data && data.last_seq && data.last_seq !== self.since) {
+            self.since = data.last_seq
+            self.ee.emit(EVENT_SEQ, self.since)
+          }
+
+          // stop on empty batch or small batch
+          if (self.stopOnEmptyChanges && data && typeof data.results !== 'undefined' && data.results.length < self.batchSize) {
+            self.continue = false
+          }
+
+          // if we have data
+          if (data && data.results && data.results.length > 0) {
+            // emit EVENT_CHANGE events
+            for (const i in data.results) {
+              self.ee.emit(EVENT_CHANGE, data.results[i])
+            }
+
+            // batch event
+            // emit EVENT_BATCH event
+            if (self.wait) {
+              // in 'wait' mode, we need to wait until the user calls
+              // a 'done' function before issuing the next change request
+              await new Promise((resolve, reject) => {
+                self.ee.emit(EVENT_BATCH, data.results, () => {
+                  resolve()
+                })
+              })
+            } else {
+              // when not in 'wait' mode, we can emit the results
+              // and continue immediately, unless there were zero results -
+              // we don't want to poll too quickly
+              self.ee.emit(EVENT_BATCH, data.results)
+            }
+          }
+        } catch (err) {
+          // error (wrong password, bad since value etc)
+          err.statusCode = err.statusCode || 500
```

Comment:
Why patching local errors?

Any error caused by a local bug or mishandling of the data will be hidden by this line. Let's instead allow such local exceptions to be bubbled up without any modification.

Reply:
👍